### PR TITLE
NIRISS AMI simulation and pipeline/ImplaneIA calibration notebooks

### PR DESCRIPTION
**Pull request by dthatte**
1_niriss_ami_binary.ipynb simulates data for binary point source AB Dor and calibrator HD37093 using Mirage. 2_niriss_ami_binary.ipynb calibrates the data with JWST pipeline and then reduces the calibrated data to fringe observables  using ImPlaneIA. The first notebook is complete and ready for review. The second notebook is work in progress.

### CODE REVIEW

**Comment by ibusko** [comment]
## Technical review of notebook #1 (1_niriss_ami_binary.ipynb)

### Running the notebook

 - The notebook runs to completion, given the requirement specs provided in the requirements files.

 - It seems that a separate pre-requirements file is also necessary, in order to force the installation of 'numpy' before everything else is installed.

 - There is apparently a bug in the mirage suite that causes calls to matplotlib functions to stall, once they are run after a run of mirage. A temporary fix was added to the matplotlib cell that stalls (#11). This should go away eventually with newer versions of mirage.

### Input data

  - Are properly stored in Box using redirect.

  - The notebook also uses pipeline reference files that can be accessed directly only from STScI's internal nodes. External users have to separately download the files; the notebook provides a pointer to the instructions.

### Libraries

 - Currently available core-libraries functionalities and visualization tools are used.

### PEP-8 and style compliance:

  - The notebook complies with the required stylistic conventions.
 
  - Running **flake8** on the notebook cells results in this report

  cell 4 line 3: E117 over-indented
  cell 6 lines 1-3: E265 block comment should start with '# '
  cell 9 line 6: E231 missing whitespace after ':'
  cell 10 line 1: E265 block comment should start with '# '
  cell 11 line 1-6: E251 unexpected spaces around keyword / parameter equals
  cell 11 line 11: E265 block comment should start with '# '
  cell 12 lines 14-15: E303 too many blank lines (2)
  cell 12 lines 14-15: E303 too many blank lines (2)
  cell 12 lines 17: E111 indentation is not a multiple of four
  cell 12 line 19: E111 indentation is not a multiple of four
  cell 12 lines 20-21: E303 too many blank lines (2)
  cell 12 lines 25-26: E303 too many blank lines (2)
  cell 12 line 29: E265 block comment should start with '# '
  cell 13 lines 11-12: E265 block comment should start with '# '
  cell 13 line 16: E251 unexpected spaces around keyword / parameter equals
  cell 13 line 23: E251 unexpected spaces around keyword / parameter equals
  cell 14 line 3: E265 block comment should start with '# '
  cell 14 lines 5: E251 unexpected spaces around keyword / parameter equals

### Build

 - There is a CI build error, but caused by an unrelated notebook (NGC4151_FeII_ContinuumFit.ipynb)


### Developer Notes

 - In the Introduction section, the developer note "Activate your mirage environment..." sounds a bit redundant, since in order to just start running the notebook, the environment must be activated already. Although, I would leave it in there for now, since it might be useful for people that are reading the static GitHub rendering of the notebook.

### Comments

 - Right after the 'Imports' section heading, the line that reads "Describe the libraries we're using...." can be removed. It sounds more like a line coming from the template notebook.

 - Cell #3 (Loading input files) could benefit from a couple of **print** statements to alert the user of what is happening under the hood. Once the zip file is downloaded that wouldn't be necessary, but when the notebook is run for the first time, the download can take some time to finish and the user may wonder if something is broken.

 - Typo: in cell #7, the second installment of the string "PAV3 for observation 1", should read instead "PAV3 for observation 2".

File: jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+    "pav3_obs1 = yaml_generator.default_obs_v3pa_on_date(pointing_name, obs1, date='2022-04-01')\n",
+    "obs2 = 2\n",
+    "pav3_obs2 = yaml_generator.default_obs_v3pa_on_date(pointing_name, obs2, date='2022-04-01')\n",
+    "\n",
+    "roll_angle = {'001':pav3_obs1 , '002': pav3_obs2}\n",
+    "\n",
+    "dates = '2022-04-01'\n",
+    "reffile_defaults = 'crds'\n",
+    "datatype = 'raw'\n",
+    "print(\"PAV3 for observation 1\", pav3_obs1, \"PAV3 for observation 1\", pav3_obs2)"
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb:290]: The second instance of "PAV3 for observation 1" should be changed to "PAV3 for observation 2"

File: jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+    "seed_images = sorted(glob.glob('mirage_sim_data/jw*final_seed_image.fits'))\n",
+    "for i, df in enumerate(seed_images):\n",
+    "    seed_im = fits.open(df)\n",
+    "    seed_im.info()\n",
+    "    im = seed_im[1].data    \n",
+    "    print(im.shape)\n",
+    "    data_seed.append(im)\n",
+    "print(data_seed[0].shape, data_seed[1].shape)\n",
+    "\n",
+    "#Temporarily adding the two lines below. \n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb:440]: These lines can be removed if you switch the order of the imports at the top of the notebook. matplotlib statements should be last in that cell, after Mirage.

File: jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+    "import sys\n",
+    "\n",
+    "import numpy as np\n",
+    "from astropy.io import fits\n",
+    "import yaml\n",
+    "import zipfile\n",
+    "import urllib.request\n",
+    "from IPython.display import Image\n",
+    "\n",
+    "%matplotlib inline\n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/1_niriss_ami_binary.ipynb:68]: I was able to get plt.imshow to work correctly when I moved these two lines below the mirage import lines (and then deleted the matplotlib lines in the cell containing plt.imshow)

**Review by bhilbert4 — APPROVED** [review]
Looks great! Just one typo, and reordering the import statements are the only issue I found.

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+   "source": [
+    "## Imports\n",
+    "Describe the libraries we're using here. If there's something unusual, explain what the library is, and why we need it.\n",
+    "- *numpy* to handle array functions\n",
+    "- *astropy.io fits* for accessing FITS files\n",
+    "- *matplotlib.pyplot* for plotting data\n",
+    "- *zipfile* for accessing zip file\n",
+    "- *urllib.request* to access URL\n",
+    "- *jwst.pipeline Detector1Pipeline, Image2Pipeline* for calibrating raw data\n",
+    "- ImplaneIA to extract ingterferometric obssrvables from calibrated data"
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:42]: typo here: "ingterferometric"

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "## Loading data\n",
+    "Upload simulated data created by 1_niriss_ami_binary.ipynb and required reference files to calibrate data without bad pixels."
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:130]: Do you mean "Download" here?

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Define output directory to save pipeline output products and default configuration files.\n",
+    "odir = './pipeline_calibrated_data/'\n",
+    "if not os.path.exists(odir):\n",
+    "    os.mkdir(odir)\n",
+    "\n",
+    "# Define path to non-default referencde file.\n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:223]: typo: "referencde"

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+    "# We are using these files to calibrate data that is simukated without bad pixels.\n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:224]: typo: "simukated"

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mirdatafiles = ['jw01093001001_01101_00001_nis_calints.fits',\n",
+    "                'jw01093002001_01101_00001_nis_calints.fits',\n",
+    "                   ]\n",
+    "\n",
+    "FIRSTFEW = None # eg 5, or None to analyze all slices\n",
+    "FIRSTFEW = 5\n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:509]: FIRSTFEW is defined on two consecutive lines here.

File: jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+    "import time\n",
+    "\n",
+    "import numpy as np\n",
+    "from astropy.io import fits\n",
+    "import matplotlib.pyplot as plt\n",
+    "from astropy import units as u\n",
+    "import zipfile\n",
+    "import urllib.request\n",
+    "\n",
+    "from jwst.pipeline.collect_pipeline_cfgs import collect_pipeline_cfgs\n",
```
Comment (bhilbert4) [inline jdat_notebooks/niriss_ami_binary/2_niriss_ami_binary.ipynb:68]: collect_pipeline_cfgs is imported, but never used.

**Review by bhilbert4 — APPROVED** [review]
This notebook also looks great! Just a few typos.

**Comment by ibusko** [comment]
## Follow up on technical review of notebook #1 (1_niriss_ami_binary.ipynb)

The issues pointed out in the first review were solved, with the exception of:

- the matplotlib import and inline commands in cell 11 are no longer necessary. They can be removed, along with the comment lines.

### PEP-8 compliance:

  - Running 'flake8' on the notebook cells results in this report:

cell 9 lines 1-6: E251 unexpected spaces around keyword / parameter equals
cell 11 lines 16,19,23: E251 unexpected spaces around keyword / parameter equals
cell 13 lines 5: E251 unexpected spaces around keyword / parameter equals
cell 15 lines 9,12,15: E251 unexpected spaces around keyword / parameter equals

**Comment by ibusko** [comment]
## Technical review of notebook #2 (2_niriss_ami_binary.ipynb)

### Running the notebook

 - The notebook runs to completion, given the requirement specs provided in the requirements files. For compatibility with notebook #1, the environment should be built under python 3.6.

 - A separate pre-requirements file is necessary, in order to force the installation of 'numpy' and 'astropy' before everything else is installed. 

### Input data

  - Are properly stored in Box using redirect.

  - The notebook uses reference files that are accessed via path defined by environmental variable WEBBPSF_PATH. This variable must be initialized before the Jupiter server starts. It is not clear how external users should set up that path. The notebook should note this, and provide instructions on how to define the path. File environment.sh contains the path currently being used.

### Libraries

 - Currently available core-libraries functionalities and visualization tools are used.

### PEP-8 compliance:

  - Running 'flake8' on the notebook cells results in this report:

  Cell 5 line 1: E265 block comment should start with '# '
  Cell 6 lines 9,12,15: E251 unexpected spaces around keyword /
  Cell 7 lines 1,6,7,10: E265 block comment should start with '# '
  Cell 8 line 4: E265 block comment should start with '# '
  Cell 8 line 4: E115 expected an indented block (comment)
  Cell 8 lines 15,25: E231 missing whitespace after ','
  Cell 12 line 7: E265 block comment should start with '# '
  Cell 12 lines 8,11,14: E251 unexpected spaces around keyword / parameter equals
  Cell 13 lines 1,3,19,28-36,53,54,59-64,69,75,82,87-90: E251 unexpected spaces around keyword / parameter equals
  Cell 13 lines 8,12,29-36,59-64,87-90: E128 continuation line under-indented for visual indent
  Cell 13 line 1: E302 expected 2 blank lines, found 1
  Cell 13 lines 8,12,22,29: E231 missing whitespace after ','
  Cell 13 line 23: E201 whitespace after '('
  Cell 13 lines 46,47,51,70,83,84: E701 multiple statements on one line (colon)
  Cell 13 line 58: E303 too many blank lines (2)
  Cell 14 line 3: E124 closing bracket does not match visual indentation
  Cell 14 lines 14,20: E231 missing whitespace after ', 24,'
  Cell 14 lines 24-28: E127 continuation line over-indented for visual indent
  Cell 14 lines 14,16,17,21-28: E251 unexpected spaces around keyword / parameter equals
  Cell 15 line 1: E265 block comment should start with '# '
  Cell 18 lines 1,4,7,10: E251 unexpected spaces around keyword / parameter equals
  Cell 18 line 1: E231 missing whitespace after ','
  Cell 19 line 2: E251 unexpected spaces around keyword / parameter equals
  Cell 21 lines 2,3: E231 missing whitespace after ','
  Cell 21 line7: E251 unexpected spaces around keyword / parameter equals


### Build

 - There is a CI build error, but caused by an unrelated notebook (Spectral Extraction.ipynb)


### Developer Notes

None


### Comments

 - Like in notebook #1, cell #1 (the one with the import statements) could be followed by a note similar to the one in notebook #1 about PYSIAF upgrades (in case it applies to this notebook, of course).

 - Right after the 'Imports' section heading, the line that reads "Describe the libraries we're using...." can be removed. It sounds more like a line coming from the template notebook.

 - Cell #1 should be broken in two: the first should have the imports and directives for everything *but* matplotlib. The cell immediately following should contain the imports and directives related to matplotlib. This is a pattern being used in several other notebooks, which helps in minimizing Matplotlib-related problems like the one found during notebook 1's development.  

 - In cell #2, the commented-out import statement should be removed.

**Comment by ibusko** [comment]
The CI build error previously found in this PR was fixed with the last commit above (f02a278). The build error that we get now, is unrelated to this directory (it is associated with directory mos-spectroscopy).

The PR is good to be merged.

**Review by ibusko — APPROVED** [review]
It's good to be merged, from the technical standpoint.

**Pull request merged**